### PR TITLE
ENH: IO support for R data files with C extension

### PR DESCRIPTION
**Pull request by ParfaitG**
- [X] follows up on #40884
- [X] tests added / passed
- [X] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [X] whatsnew entry
- [X] user_guide/io entry

### Proposed RData I/O module interfaces to the C library: [librdata](https://github.com/WizardMac/librdata). 

Overall, this PR includes following changes:
- **setup.py**: pandas/setup.py (new rdata section at the bottom)
- **librdata**: pandas/_libs/src/librdata (C and header files and iconv scripts)
- **rdata IO**: pandas/io/rdata (Cython and Python scripts)
- **frame.py**: pandas/core/frame.py (for `DataFrame.to_rdata`)
- **tests**: pandas/tests/io/test_rdata.py
- **tests data**: pandas/tests/io/data/rdata (R data files in gzip compression)
- **docs**: librdata license, user_guide/io.rst, whatsnew/v1.3.0.rst

Note: special handling of `iconv`, a system resource built-in to Unix machines, is required:
- For Linux, to centralize from different locations, the `iconv.h` header of the GNU C library is included.
- For Mac, setup.py points to system folders, `/usr/include` and `/usr/lib`, which may differ from users installs.
- For Windows , since `iconv` is not built-in, two counterpart files (.h and .c script) were added from this repo, [win-iconv](https://github.com/win-iconv/win-iconv), where its readme indicates code is placed in the public domain.

### CODE REVIEW

**Comment by ParfaitG** [comment]
@jreback, @bashtage - What are your thoughts of this C extension? Biggest challenge was handling `iconv` across OSes.

File: pandas/io/rdata/__init__.py [hunk 1]
```diff
@@ -0,0 +1 @@
+from pandas.io.rdata.rdata_reader import read_rdata  # noqa
```
Comment (ShaharNaveh) [inline pandas/io/rdata/__init__.py:1]: Can you remove the ```# noqa```, and have a ```__all__ = ["read_rdata"]```?

File: doc/source/user_guide/io.rst [hunk 2]
```diff
@@ -5906,6 +5907,307 @@ respective functions from ``pandas-gbq``.
...
+   In [608]: ghcran = pd.read_rdata("s3://public-r-data/ghcran.Rdata")
+   ...
+   UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe9 in position 45: invalid continuation byte
+
+Also, remember if R data files do not contain any data frame object, a parsing error
+will occur:
+
+.. code-block:: ipython
+
+   In [608]: rds_file = os.path.join(file_path, "env_data_non_dfs.rda")
```
Comment (bashtage) [inline doc/source/user_guide/io.rst:6058]: This seems to be missing a call to read_rds

File: doc/source/user_guide/io.rst [hunk 3]
```diff
@@ -5906,6 +5907,307 @@ respective functions from ``pandas-gbq``.
...
+   plants_df = pd.read_rdata(rds_file, rownames=False)["r_dataframe"].tail()
+   plants_df
+
+
+To select specific objects in .rda, pass a list of names into ``select_frames``:
+
+.. ipython:: python
+
+   rda_file = os.path.join(file_path, "env_data_dfs.rda")
+   env_dfs = pd.read_rdata(rda_file, select_frames=["sea_ice_df"])
```
Comment (bashtage) [inline doc/source/user_guide/io.rst:6018]: Is this a good pattern?  Should it default to 

(a) reading a single frame if there is one
(b) raising and telling the user that a specific frame name is required?

or

(a) reading a dingle frame if available
(b) reading the "first" frame where first is based on either the order in the rdata file or alphabetically

Think about how this compares to `read_excel` which supports reading multiple frames from the same file?

In fact, should there be a class-based interface that would allow multiple reads from the same file, similar to `ExcelReader`?

File: pandas/_libs/src/librdata/win_iconv.c [hunk 4]
```diff
@@ -0,0 +1,2228 @@
+/*
+ * No Copyright.
```
Comment (bashtage) [inline pandas/_libs/src/librdata/win_iconv.c:2]: Perhaps license CC0 is cleaner?

File: pandas/core/frame.py [hunk 5]
```diff
@@ -2295,6 +2295,138 @@ def _from_arrays(
         )
         return cls(mgr)
 
+    @doc(storage_options=generic._shared_docs["storage_options"])
+    def to_rdata(
+        self,
+        path_or_buffer: FilePathOrBuffer,
+        file_format: str = "infer",
+        rda_name: str = "pandas_dataframe",
```
Comment (bashtage) [inline pandas/core/frame.py:2303]: Something closer to what `read_excel` does where ti chooses sheets[0] by default probably makes more sense.

File: pandas/core/frame.py [hunk 6]
```diff
@@ -2295,6 +2295,138 @@ def _from_arrays(
...
+            DataFrame indexes can include multiple columns and R rownames can
+            only include one column, DataFrame index will not map to R
+            data.frame rownames.
+
+        compression : {{'gzip', 'bz2', 'xz', None}}, default 'gzip'
+            Compression type for on-the-fly decompression of on-disk data.
+
+        {storage_options}
+
+        Returns
```
Comment (bashtage) [inline pandas/core/frame.py:2338]: I think you can drop the Returns section here.  You coudl add a Raises section though.

File: pandas/io/rdata/_rdata.pyx [hunk 7]
```diff
@@ -0,0 +1,437 @@
...
+        if vtype == "object":
+            for k, v in vdata.items():
+                if v == v:
+                    rdata_append_string_value(self.writer, v)
+                else:
+                    rdata_append_string_value(self.writer, NULL)
+
+        rdata_end_column(self.writer, py_col)
+
+    cpdef write_rdata(self, rfile, rdict, rformat, tbl_name=None):
```
Comment (bashtage) [inline pandas/io/rdata/_rdata.pyx:362]: Why not plain `def`?

File: pandas/io/rdata/_rdata.pyx [hunk 8]
```diff
@@ -0,0 +1,437 @@
...
+        bint is_factor
+        dict rownames
+        dict colnames
+        dict rtypes
+        str tblname
+        dict rvalues
+        int dims
+        dict dim_str
+
+    cpdef read_rdata(self, rfile):
```
Comment (bashtage) [inline pandas/io/rdata/_rdata.pyx:245]: Plain `def` should be fine here too, since this isn't performance sensitive code that needs to be called by Cython.

File: pandas/tests/io/test_rdata.py [hunk 9]
```diff
@@ -0,0 +1,565 @@
...
+        read_rdata(url)
+
+
+# S3
+
+
+@tm.network
+@td.skip_if_no("s3fs")
+def test_read_rda_s3():
+    s3 = "s3://assets.datacamp.com/production/course_1478/datasets/wine.RData"
```
Comment (bashtage) [inline pandas/tests/io/test_rdata.py:301]: Is this a good idea?  Could trigger CI failures if datacamp makes changes.

File: pandas/tests/io/test_rdata.py [hunk 10]
```diff
@@ -0,0 +1,565 @@
+from io import BytesIO
```
Comment (bashtage) [inline pandas/tests/io/test_rdata.py:1]: Need to add a test where a reader and a writer  is pickled and unpickled to make sure the Cython classes work as expected aroudn pickling.

File: pandas/tests/io/test_rdata.py [hunk 11]
```diff
@@ -0,0 +1,565 @@
...
+import os
```
Comment (bashtage) [inline pandas/tests/io/test_rdata.py:2]: Can you add a test that writes and then rereads a DF that contains every non-object dtype. What about ExtensionTypes?

File: pandas/tests/io/test_rdata.py [hunk 12]
```diff
@@ -0,0 +1,565 @@
...
+        },
+        "region": {1012: "S", 1013: "S", 1014: "S", 1015: "S", 1016: "S"},
+        "extent": {1012: 8.28, 1013: 9.48, 1014: 9.19, 1015: 9.41, 1016: 10.44},
+        "area": {1012: 5.51, 1013: 6.23, 1014: 5.59, 1015: 6.59, 1016: 6.5},
+    }
+).rename_axis("rownames")
+
+ppm_df = DataFrame(
+    {
+        "date": {
```
Comment (bashtage) [inline pandas/tests/io/test_rdata.py:80]: Can R data files have dates outside the date range supported by pandas?  What happens there? Can you add a test for this case?

File: setup.py [hunk 13]
```diff
@@ -629,6 +637,49 @@ def srcpath(name=None, suffix=".pyx", subdir="src"):
...
+
+if is_platform_windows():
+    rdata_srcs.append("pandas/_libs/src/librdata/win_iconv.c")
+
+include = []
+libs_dir = []
+libs = []
+if is_platform_mac():
+    os.environ["DYLD_LIBRARY_PATH"] = ""
+    include = ["/usr/include"]
```
Comment (bashtage) [inline setup.py:662]: Why not handle the mac includes here rathe than up top?

File: pandas/io/rdata/__init__.py (hunk 1 above)
Comment (ParfaitG) [reply pandas/io/rdata/__init__.py:1]: Good suggestion. Will change. I followed `pandas.io.sas` here.

File: doc/source/user_guide/io.rst (hunk 2 above)
Comment (ParfaitG) [reply doc/source/user_guide/io.rst:6058]: Good catch. Will fix.

File: doc/source/user_guide/io.rst (hunk 3 above)
Comment (ParfaitG) [reply doc/source/user_guide/io.rst:6018]: Originally, I had output generate either a single DataFrame for `.rds` types or dict of DataFrames for `.RData`/`.rda` types. So the file format type decided the structure. Earlier, jreback was confused by this, so now I simplified to always return a dict of DataFrames for _any_ R data file type. Because rds types carry no named objects unlike rda types, the default `r_dataframe` is used as key for single-item dict. But rda object names will map into dict keys.

For vast majority, especially in R packages, only a single named data frame is ever stored in .Rdata or .rda files. Given rdata IO may not be as popular as the Excel IO module, for class-based reader interface, maybe we should gauge this need by users for a potential future PR?

File: pandas/_libs/src/librdata/win_iconv.c (hunk 4 above)
Comment (ParfaitG) [reply pandas/_libs/src/librdata/win_iconv.c:2]: How about this using [CC0 wiki format](https://wiki.creativecommons.org/wiki/CC0_FAQ#May_I_apply_CC0_to_computer_software.3F_If_so.2C_is_there_a_recommended_implementation.3F)? May need `copyright` somewhere to pass `cpplint`.

> win-iconv - iconv implementation using Win32 API to convert.
Written in 2009 by Yukihiro Nakadaira (ynkdir) <<https://github.com/ynkdir>> and 
> Contributors to win-iconv <<https://github.com/win-iconv/win-iconv>>
>
> To the extent possible under law, the author(s) have dedicated all copyright and related and neighboring rights to this software to the public domain worldwide. This software is distributed without any warranty.
You should have received a copy of the CC0 Public Domain Dedication along with this software. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.

File: pandas/core/frame.py (hunk 5 above)
Comment (ParfaitG) [reply pandas/core/frame.py:2303]: Not clear what you mean here. Currently, librdata supports only writing a single `DataFrame`  to either R data file format. Also, `read_rdata` (not shown here in `frame.py`) maps R object names to dict keys.

File: pandas/core/frame.py (hunk 6 above)
Comment (ParfaitG) [reply pandas/core/frame.py:2338]: Agreed. Will adjust.

File: pandas/core/frame.py [hunk 14]
```diff
@@ -2295,6 +2295,138 @@ def _from_arrays(
...
+            self,
+            path_or_buffer=path_or_buffer,
+            file_format=file_format,
+            rda_name=rda_name,
+            index=index,
+            compression=compression,
+            storage_options=storage_options,
+        )
+
+        return r.write_data()
```
[the remark this answers is not in the document]
Comment (ParfaitG) [reply pandas/core/frame.py:2428]: Agreed. Will adjust.

File: pandas/core/frame.py [hunk 15]
```diff
@@ -2295,6 +2295,138 @@ def _from_arrays(
...
+        ... 1     0 Pteridophytes      Data Deficient   398
+        ... 2     1 Pteridophytes             Extinct    65
+        ... 3     2 Pteridophytes      Not Threatened  1294
+        ... 4     3 Pteridophytes Possibly Threatened   408
+        ... 5     4 Pteridophytes          Threatened  1275
+        ... '''
+        """
+        from pandas.io.rdata.rdata_writer import RDataWriter
+
+        r = RDataWriter(
```
[the remark this answers is not in the document]
Comment (ParfaitG) [reply pandas/core/frame.py:2418]: Sounds good.

File: pandas/io/rdata/_rdata.pyx (hunk 7 above)
Comment (ParfaitG) [reply pandas/io/rdata/_rdata.pyx:362]: Learning as I was doing. Either way can work.

File: pandas/io/rdata/_rdata.pyx (hunk 8 above)
Comment (ParfaitG) [reply pandas/io/rdata/_rdata.pyx:245]: Sounds good.

File: pandas/tests/io/test_rdata.py (hunk 9 above)
Comment (ParfaitG) [reply pandas/tests/io/test_rdata.py:301]: Good point. Can't find a usable other example unless we push to pandas' S3 account.

File: pandas/tests/io/test_rdata.py (hunk 10 above)
Comment (ParfaitG) [reply pandas/tests/io/test_rdata.py:1]: Can you explain a little more here? Can you point me to a working example of such pickling?

File: pandas/tests/io/test_rdata.py (hunk 11 above)
Comment (ParfaitG) [reply pandas/tests/io/test_rdata.py:2]: Will do. I didn't explore ExtensionTypes which may need conversion handling to R data types.

File: pandas/tests/io/test_rdata.py (hunk 12 above)
Comment (ParfaitG) [reply pandas/tests/io/test_rdata.py:80]: What date range would that be?

File: setup.py (hunk 13 above)
Comment (ParfaitG) [reply setup.py:662]: Didn't want to interfere with above extensions especially with modification of environment variable in session.

File: pandas/tests/io/test_rdata.py (hunk 12 above)
Comment (bashtage) [reply pandas/tests/io/test_rdata.py:80]: Pandas only supports dates in +/- 293 years from 1970.
```
In [29]: pd.Timestamp.min
Out[29]: Timestamp('1677-09-21 00:12:43.145225')

In [30]: pd.Timestamp.max
Out[30]: Timestamp('2262-04-11 23:47:16.854775807')
```

File: pandas/tests/io/test_rdata.py (hunk 10 above)
Comment (bashtage) [reply pandas/tests/io/test_rdata.py:1]: Pickling of non-trivial cython classes can require an implementation of __setstate__ and __reduce__.  You can write the test to see if it works.

1. Create Reader or Writer
2. Pickle to string using dumps and reload using loads
3. Use the unpickled Reader or Writer to read or write a file.

If the test passes, and the unpickled reader/writer perform the same as a fresh instance, then you are OK.

File: pandas/core/frame.py (hunk 5 above)
Comment (bashtage) [reply pandas/core/frame.py:2303]: All readers (AFAICT) pull in a single dataframe, including read_excel even if the sheet has multiple copies.  Perhaps the default should be to return the first DataFrame-like object in the rdata file.  This is more-or-less what ExcelReader does.

File: doc/source/user_guide/io.rst (hunk 3 above)
Comment (bashtage) [reply doc/source/user_guide/io.rst:6018]: I think that the `read_...` method should be simple and return a `DataFrame`.  Users who need to interact with complicated R data files should be directed to use the lower-level class-based reader.  This is the model that is used in `StataReader` and IIRC in `ExcelFile`.

File: setup.py (hunk 13 above)
Comment (bashtage) [reply setup.py:662]: But this include here is _only_ used by rdata_ext, so you can customize it for rdata_ext as needed rather than injecting the customization above.

Comment (bashtage) [reply setup.py:662]: I think you should devine extra includes for this module in the special block if OSX that occurs earlier. You can then add these includes in the usual place in the dictionary and avoid all of the special casing I think.

File: pandas/tests/io/test_rdata.py (hunk 12 above)
Comment (bashtage) [reply pandas/tests/io/test_rdata.py:80]: A related question - are there any forbidden names for objects in rdata files?  For example, keywords in R?

File: doc/source/user_guide/io.rst (hunk 3 above)
Comment (ParfaitG) [reply doc/source/user_guide/io.rst:6018]: Hmmm...per IO docs under [specifying sheets](https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html#specifying-sheets), if you pass a list of sheets or `None`, `read_excel` returns a dict of DataFrames without specifying `ExcelFile`. This is the model used here for `read_rdata`. Also, [`read_html`](https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html#io-read-html) only returns a list of DataFrames. I thought StataReader and SASReader were more for reading files incrementally with chunks and any meta data.

File: pandas/tests/io/test_rdata.py (hunk 12 above)
Comment (ParfaitG) [reply pandas/tests/io/test_rdata.py:93]: R has a handful [reserved words](https://stat.ethz.ch/R-manual/R-devel/library/base/html/Reserved.html) that are not allowed for object names unless escaped with backticks.

File: doc/source/user_guide/io.rst [hunk 16]
```diff
@@ -5906,6 +5907,370 @@ respective functions from ``pandas-gbq``.
...
+.. _rds: https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/readRDS
+
+.. _rda: https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/save
+
+For example, consider the following generated data.frames in R using environment
+data samples from US EPA, UK BGCI, and NOAA pubilc data:
+
+.. code-block:: r
+
+   ghg_df <- data.frame(
```
Comment (jreback) [inline doc/source/user_guide/io.rst:5947]: is it possible to show the r dataframe here (hard-code ok), just to get a view of hit on the r side)

File: doc/source/user_guide/io.rst (hunk 3 above)
Comment (jreback) [reply doc/source/user_guide/io.rst:6018]: yeah for `read_hdf` we do this by forcing the user to have a key. are these always an ordered list? or a keyed list?

**Review by jreback — CHANGES_REQUESTED** [review]
@ParfaitG is this a different format than pyarrow reads/writes? (on the r side) sorry it obviously is, what i mean is, are there utilities / code in the pyarrow r project to read write this format?

**Comment by ParfaitG** [comment]
@jreback - These binary formats (RData, rda, rds) are part of `base` R (or the standard library of R) as native serialization types. Methods to create them ship with every installation of R. The R Core Team maintain these types. AFAIK, while the `arrow` package of R and `pyarrow` of Python can read text files like csv and json and compress to gz and bzip2 types, they do not read or write to these R data formats. From what I see, the only binary types Arrow supports are `parquet` and `feather`. And from McKinney and Wickham blogs, I believe these formats serve as alternatives for their speed in read/write and language-independent interoperability.

File: doc/source/user_guide/io.rst (hunk 16 above)
Comment (ParfaitG) [reply doc/source/user_guide/io.rst:5947]: Yep. No problem. Can do.

**Comment by ParfaitG** [comment]
What is the status here? I see no further response. I can rebase but this ENH may not make it into v1.3. I fully understand that this PR with new C extension is pretty involved and we want to be conservative on large enhancements in `pandas` code base.

**Comment by jreback** [comment]
@ParfaitG have been busy but will take a look

File: setup.py [hunk 17]
```diff
@@ -570,6 +582,13 @@ def srcpath(name=None, suffix=".pyx", subdir="src"):
     include = data.get("include", [])
     include.append(numpy.get_include())
 
+    if name == "io.rdata._rdata" and is_platform_mac():
+        # non-conda builds must adjust paths to libiconv .h and lib dirs
+        include = [
+            os.path.join(os.environ["CONDA_PREFIX"], "include"),
```
Comment (bashtage) [inline setup.py:588]: Should you check if CONDA_PREFIX is defined?  If it isn't, presumably this isn't conda.

File: setup.py [hunk 18]
```diff
@@ -364,6 +370,12 @@ def run(self):
     # https://github.com/pandas-dev/pandas/issues/35559
     extra_compile_args.append("-Wno-error=unreachable-code")
 
+    # rdata requires system iconv library
+    os.environ["DYLD_LIBRARY_PATH"] = ""
```
Comment (bashtage) [inline setup.py:374]: Overriding the users DYLC_LIBRARY_PATH feels like the wrong solution.  Why is this needed? Does this change this path for the duration of the console session, even outside of the setup run?

File: setup.py [hunk 19]
```diff
@@ -364,6 +370,12 @@ def run(self):
...
+    rdata_includes = ["/usr/include"]
```
Comment (bashtage) [inline setup.py:375]: Presumably these should be no on Windows

**Review by bashtage — COMMENTED** [review]
Some comments on setup.

**Review by jreback — CHANGES_REQUESTED** [review]
is it possible to push all of the c-code to a separate package? why is this a good (or not good solution in this case)? i am concerned that this is just creating even bigger wheels and more maintenance for all users.

File: setup.py (hunk 17 above)
Comment (ParfaitG) [reply setup.py]: Thanks for this catch. Turns out this `if` block is not needed. I removed it.

File: setup.py (hunk 18 above)
Comment (ParfaitG) [reply setup.py]: This line is not needed and has been removed.

File: setup.py (hunk 19 above)
Comment (ParfaitG) [reply setup.py:374]: Correct. This section is under the `if is_platform_mac()` condition.

**Comment by ParfaitG** [comment]
If you recall, we explored a Python package for this IO module to read R data files but it had a restrictive license. See previous PR in my first checked item at top. That package was a wrapper to this same C library that we directly interfaced with here, avoiding another soft dependency for pandas.

Regarding size, this [librdata](https://github.com/WizardMac/librdata) proclaims itself as a small, lightweight C library. Its source files with two added scripts for Windows totals 177 KB (slightly larger than `ujson` at 175 KB). Also, the *only* compiled Cython `_rdata` module totals 186 KB for Python 3.9 on Linux, likely about the same for Py 3.7/3.8. By comparison, SAS's compiled Cython module totals 256 KB. Corresponding `_rdata.c` comes to 533 KB (with SAS's _sas.c at 1.0 MB).
